File: crates/adapters/inbound/powehi-rest-api/src/routes/groups.rs
//! Group management routes.
//!
//! Clients register their MLS group with the server before sending messages.
//! The creator becomes the first group member, enabling the fail-closed
//! membership check in `MessagingService`.
//!
//! Member add/remove are gated on caller membership (fail-closed): only an
//! existing member may add or remove another device.  This mirrors the
//! application-layer check in `GroupService` and prevents hijacking by
//! authenticated-but-non-member devices.

use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};
use powehi_domain::{
    device::DeviceId,
    group::{Epoch, GroupId},
};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{error::ApiError, middleware::AuthenticatedDevice, AppState};

#[derive(Deserialize)]
pub struct CreateGroupRequest {
    pub group_id: GroupId,
}

pub async fn create_group(
    State(state): State<AppState>,
    AuthenticatedDevice(creator): AuthenticatedDevice,
    Json(req): Json<CreateGroupRequest>,
) -> Result<StatusCode, ApiError> {
    tracing::info!(
        creator = %creator,
        group_id = %req.group_id,
        "groups.create_group"
    );
    state.group.create_group(&creator, req.group_id).await?;
    Ok(StatusCode::NO_CONTENT)
}

/// Request body for adding a group member.
///
/// `epoch` is the MLS epoch at which the new member joined.  It is stored by
/// the server as opaque metadata for potential future use (e.g. audit, fan-out
/// scoping).  The server does NOT validate the epoch value; MLS semantics are
/// enforced end-to-end by the clients.
#[derive(Deserialize)]
pub struct AddMemberRequest {
    pub epoch: u64,
}

/// `POST /v1/groups/:group_id/members/:device_id`
///
/// Registers `device_id` as a member of `group_id` at `epoch`.  The caller
/// (authenticated device) must already be a member of the group; the
/// application layer returns `Unauthorized` otherwise.
pub async fn add_member(
    State(state): State<AppState>,
    AuthenticatedDevice(caller): AuthenticatedDevice,
    Path((raw_group_id, raw_device_id)): Path<(Uuid, Uuid)>,
    Json(req): Json<AddMemberRequest>,
) -> Result<StatusCode, ApiError> {
    let group_id = GroupId::from(raw_group_id);
    let target_device_id = DeviceId::from(raw_device_id);
    tracing::info!(
        caller = %caller,
        group_id = %group_id,
        "groups.add_member"
    );
    state
        .group
        .add_member(&caller, &group_id, &target_device_id, Epoch(req.epoch))
        .await?;
    Ok(StatusCode::NO_CONTENT)
}

/// `DELETE /v1/groups/:group_id/members/:device_id`
///
/// Removes `device_id` from `group_id`.  The caller must already be a member
/// of the group; the application layer returns `Unauthorized` otherwise.
pub async fn remove_member(
    State(state): State<AppState>,
    AuthenticatedDevice(caller): AuthenticatedDevice,
    Path((raw_group_id, raw_device_id)): Path<(Uuid, Uuid)>,
) -> Result<StatusCode, ApiError> {
    let group_id = GroupId::from(raw_group_id);
    let target_device_id = DeviceId::from(raw_device_id);
    tracing::info!(
        caller = %caller,
        group_id = %group_id,
        "groups.remove_member"
    );
    state
        .group
        .remove_member(&caller, &group_id, &target_device_id, Epoch(0))
        .await?;
    Ok(StatusCode::NO_CONTENT)
}

/// Response body for `GET /v1/groups/:group_id/pending-removals`.
///
/// Device UUIDs only — routing metadata, never key material or ciphertext.
#[derive(Serialize)]
pub struct PendingRemovalsResponse {
    pub device_ids: Vec<DeviceId>,
}

/// `GET /v1/groups/:group_id/pending-removals`
///
/// Returns the devices whose MLS Remove the group's remaining members still
/// owe. The caller must already be a member of `group_id`; the application
/// layer returns `Unauthorized` otherwise, which surfaces as `401
/// Unauthorized` via `ApiError`. The server can never perform the Remove
/// itself (no group state, no keys — prd.md §5.4, §8.5), so this endpoint
/// exists purely so clients can discover the work they must do.
pub async fn list_pending_removals(
    State(state): State<AppState>,
    AuthenticatedDevice(caller): AuthenticatedDevice,
    Path(raw_group_id): Path<Uuid>,
) -> Result<Json<PendingRemovalsResponse>, ApiError> {
    let group_id = GroupId::from(raw_group_id);
    tracing::info!(
        caller = %caller,
        group_id = %group_id,
        "groups.list_pending_removals"
    );
    let device_ids = state
        .group
        .list_pending_removals(&caller, &group_id)
        .await?;
    Ok(Json(PendingRemovalsResponse { device_ids }))
}

/// Maximum number of device ids returned by
/// `GET /v1/groups/:group_id/members` in one response.
///
/// Group membership itself is uncapped in this codebase
/// (`MAX_FAN_OUT_RECIPIENTS` bounds push fan-out, not membership — see
/// `powehi_application::messaging_service`), so without a cap here a single
/// authenticated member could pull O(group size) egress per request at the
/// api_governor's sustained rate. Aligned with `MAX_FAN_OUT_RECIPIENTS`
/// (512) for consistency with the other group-size amplification cap.
/// security-auditor finding, cycle 454.
pub(crate) const MAX_MEMBERS_RESPONSE: usize = 512;

/// Response body for `GET /v1/groups/:group_id/members`.
///
/// Device UUIDs only — routing metadata, never key material or ciphertext.
/// `joined_at_epoch` is deliberately NOT exposed: the domain layer carries
/// it for future use, but no client needs it and it is one more piece of
/// metadata surface (prd.md §3.3 minimalism, P5).
///
/// `device_ids` is sorted by device UUID, NOT by join order. The repository
/// returns rows `ORDER BY joined_at_epoch ASC`; serving that order would
/// leak a monotone function of the very field this response drops, so the
/// handler re-sorts into a canonical, information-free order. It also makes
/// truncation deterministic and stable across calls.
///
/// This "information-free" property holds ONLY because `DeviceId` wraps a
/// random UUIDv4 (`DeviceId::new()` -> `Uuid::new_v4()`). If device-id
/// generation ever moves to a time-sortable UUID (e.g. UUIDv7), sorting by
/// UUID would silently become a monotone function of registration time and
/// reopen the exact join-order channel this sort was added to close.
///
/// `truncated` is `true` when the group has more than
/// `MAX_MEMBERS_RESPONSE` members and `device_ids` is therefore a PREFIX,
/// not the full membership. This flag exists because this endpoint's whole
/// purpose is letting a client reconcile its own MLS ratchet tree against
/// the server's device list (prd.md §5.4): a silently truncated list would
/// make legitimate, present devices look absent — turning the defence into
/// the exact false-eviction attack it is meant to prevent. A client MUST
/// NOT treat absence from a `truncated` response as evidence of anything.
#[derive(Serialize)]
pub struct MembersResponse {
    pub device_ids: Vec<DeviceId>,
    pub truncated: bool,
}

/// `GET /v1/groups/:group_id/members`
///
/// Returns the devices the server records as members of `group_id`. The
/// caller must already be a member; the application layer returns
/// `Unauthorized` otherwise, which surfaces as `401 Unauthorized` via
/// `ApiError` — an unknown group id is answered identically at the response
/// level, so this endpoint is not a group-existence oracle in its status,
/// body, or error code (rejection latency still scales with the real
/// group's size, a side channel shared with the sibling guards this is
/// modeled on). This is intended as one half of the local cross-check for
/// the server-reported `pending-removals` signal (prd.md §5.4): the other
/// half — exposing a device_id-to-MLS-leaf mapping from the WASM crypto
/// layer so a client can actually join this list against its own ratchet
/// tree — does not exist yet, so reconciliation is not yet possible end to
/// end. See `MembersResponse` for the ordering and truncation contract.
pub async fn list_members(
    State(state): State<AppState>,
    AuthenticatedDevice(caller): AuthenticatedDevice,
    Path(raw_group_id): Path<Uuid>,
) -> Result<Json<MembersResponse>, ApiError> {
    let group_id = GroupId::from(raw_group_id);
    tracing::info!(
        caller = %caller,
        group_id = %group_id,
        "groups.list_members"
    );
    let mut device_ids: Vec<DeviceId> = state
        .group
        .list_members(&caller, &group_id)
        .await?
        .into_iter()
        .map(|m| m.device_id)
        .collect();
    // Canonical order first, then truncate: truncation must not depend on
    // the repository's join-order sort (see `MembersResponse` doc).
    device_ids.sort_by_key(|d| d.as_uuid());
    let truncated = device_ids.len() > MAX_MEMBERS_RESPONSE;
    device_ids.truncate(MAX_MEMBERS_RESPONSE);
    Ok(Json(MembersResponse {
        device_ids,
        truncated,
    }))
}

File: app/src/api/groups.ts
/**
 * Group management HTTP client.
 *
 * Groups are MLS groups registered with the server so it can enforce
 * membership for message delivery. The server sees only opaque UUIDs.
 */

const API_BASE = "/v1";

// Strict opaque ID format: lowercase hex UUID (no path traversal, no injection)
const OPAQUE_ID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

function assertOpaqueId(id: string, name: string): void {
	if (!OPAQUE_ID_RE.test(id)) {
		throw new Error(`invalid_${name}`);
	}
}

function authHeaders(token: string): HeadersInit {
	return {
		"Content-Type": "application/json",
		Authorization: `Bearer ${token}`,
	};
}

async function throwOnError(resp: Response): Promise<void> {
	if (!resp.ok) {
		const body = (await resp.json().catch(() => ({}))) as { code?: string };
		throw new Error(body.code ?? `http_${resp.status}`);
	}
}

/** POST /v1/groups — register a new MLS group. Creator becomes first member. */
export async function createGroup(token: string, groupId: string): Promise<void> {
	assertOpaqueId(groupId, "group_id");
	const resp = await fetch(`${API_BASE}/groups`, {
		method: "POST",
		headers: authHeaders(token),
		body: JSON.stringify({ group_id: groupId }),
	});
	await throwOnError(resp);
}

/**
 * POST /v1/groups/:groupId/members/:deviceId — add a device to the group.
 * Caller must be an existing group member.
 * @param epoch MLS epoch at which the member joined.
 */
export async function addMember(
	token: string,
	groupId: string,
	deviceId: string,
	epoch: number,
): Promise<void> {
	assertOpaqueId(groupId, "group_id");
	assertOpaqueId(deviceId, "device_id");
	const resp = await fetch(
		`${API_BASE}/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(deviceId)}`,
		{
			method: "POST",
			headers: authHeaders(token),
			body: JSON.stringify({ epoch }),
		},
	);
	await throwOnError(resp);
}

/**
 * DELETE /v1/groups/:groupId/members/:deviceId — remove a device from the group.
 * Caller must be an existing group member.
 */
export async function removeMember(
	token: string,
	groupId: string,
	deviceId: string,
): Promise<void> {
	assertOpaqueId(groupId, "group_id");
	assertOpaqueId(deviceId, "device_id");
	const resp = await fetch(
		`${API_BASE}/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(deviceId)}`,
		{
			method: "DELETE",
			headers: { Authorization: `Bearer ${token}` },
		},
	);
	await throwOnError(resp);
}

/**
 * GET /v1/groups/:groupId/pending-removals — device UUIDs the group still
 * "owes" an MLS Remove for (server-side revocation bookkeeping only).
 *
 * This is a REQUEST for the client to act, never a proof: the server holds no
 * group state or keys and cannot construct or verify an MLS Remove (prd.md
 * §5.4). Callers MUST treat the returned device IDs as candidates requiring
 * explicit human confirmation before calling `removeMember` — never auto-execute.
 * Caller must already be a group member (401 otherwise).
 */
export async function listPendingRemovals(token: string, groupId: string): Promise<string[]> {
	assertOpaqueId(groupId, "group_id");
	const resp = await fetch(`${API_BASE}/groups/${encodeURIComponent(groupId)}/pending-removals`, {
		method: "GET",
		headers: { Authorization: `Bearer ${token}` },
	});
	await throwOnError(resp);
	const body = (await resp.json()) as { device_ids: string[] };
	return body.device_ids;
}

/**
 * GET /v1/groups/:groupId/members — device UUIDs the server records as
 * current members of `group_id`.
 *
 * This is one half of the local cross-check for the `pending-removals`
 * signal (prd.md §5.4): joining this list against a pending device_id can
 * catch server-side inconsistency (e.g. a pending-removal entry for a
 * device that was never a member, or one already removed), but NOT a fully
 * malicious/colluding server — the other half (binding device_ids to the
 * client's own MLS ratchet tree leaves) does not exist yet. See
 * `MembersResponse` doc comment in `groups.rs` for the full contract.
 *
 * `truncated: true` means `deviceIds` is a PREFIX, not the full membership
 * — callers MUST NOT treat absence from a truncated response as meaningful.
 * Caller must already be a group member (401 otherwise).
 */
export async function listMembers(
	token: string,
	groupId: string,
): Promise<{ deviceIds: string[]; truncated: boolean }> {
	assertOpaqueId(groupId, "group_id");
	const resp = await fetch(`${API_BASE}/groups/${encodeURIComponent(groupId)}/members`, {
		method: "GET",
		headers: { Authorization: `Bearer ${token}` },
	});
	await throwOnError(resp);
	const body = (await resp.json()) as { device_ids?: unknown; truncated?: unknown };
	// Validate rather than blindly trust the cast: a malformed/missing
	// `device_ids` must fail (the caller then skips the cross-check, same as
	// any other fetch failure) instead of silently becoming e.g. a char set
	// from `new Set(someString)`. A malformed/missing `truncated` defaults to
	// `true` (safe direction) rather than `false` — per `MembersResponse`'s
	// own contract, treating absence as meaningful is the false-eviction
	// failure mode this endpoint exists to avoid, so an unrecognized shape
	// must degrade toward "skip the cross-check", never toward "trust it".
	if (!Array.isArray(body.device_ids) || !body.device_ids.every((id) => typeof id === "string")) {
		throw new Error("invalid_members_response");
	}
	return { deviceIds: body.device_ids, truncated: body.truncated !== false };
}

/**
 * GET /v1/groups/:groupId/epoch — the server's current CAS epoch counter for
 * `groupId`, scoping data for a future `sendCommit` call.
 *
 * This is NOT the client's local MLS epoch (`mls_group.rs`'s remove-member
 * status-list item (a)): it only advances through `GroupRepository::
 * advance_epoch`'s CAS on an accepted Commit, so it legitimately diverges
 * from the local MLS epoch whenever a membership change reached the server
 * through a path that never called `sendCommit`. Treat the returned value as
 * an opaque CAS token for `sendCommit`'s `expectedEpoch`, never as a stand-in
 * for the client's own local MLS epoch tracking.
 *
 * Caller must already be a group member (401 otherwise; an unknown group id
 * answers identically, same non-oracle property as `listMembers`). A
 * non-home-region caller gets `502 region_mismatch` instead of a stale-or-
 * zero epoch (fail-closed, prd.md §3.5.1) — callers MUST treat that as a
 * distinct, retryable-elsewhere failure, not as "epoch 0".
 *
 * No production caller reads this yet — issue #2's wiring pass (item (a) in
 * `mls_group.rs`'s status list) is scoping-only until one exists. When a
 * caller does wire this into a `sendCommit` flow, it MUST call `getEpoch`,
 * drain the group's poll-loop backlog, stage, and send as work inside ONE
 * `withMlsCommitLock` acquisition (`app/src/lib/mlsCommitLock.ts`) — reading
 * the epoch outside that lock, or across two separate acquisitions, is
 * structurally stale by the time `stage` runs, even though the server's CAS
 * still fails such a stale attempt closed (`409 epoch_mismatch`) rather than
 * letting it fork the group.
 */
export async function getEpoch(token: string, groupId: string): Promise<number> {
	assertOpaqueId(groupId, "group_id");
	const resp = await fetch(`${API_BASE}/groups/${encodeURIComponent(groupId)}/epoch`, {
		method: "GET",
		headers: { Authorization: `Bearer ${token}` },
	});
	await throwOnError(resp);
	const body = (await resp.json().catch(() => ({}))) as { epoch?: unknown };
	if (typeof body.epoch !== "number" || !Number.isSafeInteger(body.epoch) || body.epoch < 0) {
		throw new Error("invalid_epoch_response");
	}
	return body.epoch;
}
